Fix form authentication in tomcat MixedAuthenticator valve

amergey opened the pull request:
It fixes "java.lang.IllegalArgumentException: Path [] does not start with a "/"
character" thrown when getServletPath() return null

and also code seems to be more consistent between negociate and form authentication

dblock commented:
I am pretty sure the `if` here was not accidental and is going to break things. The real question is what is it going to break? 

- Does the demo still work?
- Can we write tests for this?

amergey commented:
You are right, the if allowed to be redirected to error page instead of a blank page, but in this case the if logic is useful also in case of negociate authentication, so previous code was already a bit wrong. 

Yes demo is still working.

The MixedAuthenticator.authenticate is already covered by test case which is testing if authenticate is returning the right boolean value. The redirect part was not covered anyway (as it is an interaction between the browser and tomcat)

dblock commented on Source/JNA/waffle-tomcat7/src/main/java/waffle/apache/MixedAuthenticator.java:
```diff
@@ -93,16 +93,18 @@ public boolean authenticate(final Request request, final HttpServletResponse res
         } else if (negotiateCheck) {
             if (!authorizationHeader.isNull()) {
-                return this.negotiate(request, response, authorizationHeader);
+                boolean negociateResult = this.negotiate(request, response, authorizationHeader);
```
negotiateResult mispelled

dblock commented:
@hazendaz should take a look at this

dblock commented:
@amergey This changes the behavior of the authenticator. It needs tests for this use-case and a demo that didn't work before and works now.

hazendaz commented:
If the intent is to fix a null pointer, wouldn't it make more sense to just handle the null pointer without changing current logic?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Daniel Doubrovkine (dB.) @dblockdotorg <notifications@github.com>
Sent: Saturday, October 12, 2019 10:43:25 AM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


@amergey<https://github.com/amergey> This changes the behavior of the authenticator. It needs tests for this use-case and a demo that didn't work before and works now.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI4P6OWXLABXTTPSZOLQOHPA3A5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBCA45I#issuecomment-541331061>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODI3REYBQRRH4343S45DQOHPA3ANCNFSM4I7ODUCQ>.

amergey commented:
> If the intent is to fix a null pointer, wouldn't it make more sense to just handle the null pointer without changing current logic? Get Outlook for Android<https://aka.ms/ghei36>
> […](#)
> ________________________________ From: Daniel Doubrovkine (dB.) @dblockdotorg <notifications@github.com> Sent: Saturday, October 12, 2019 10:43:25 AM To: Waffle/waffle <waffle@noreply.github.com> Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com> Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800) @amergey<https://github.com/amergey> This changes the behavior of the authenticator. It needs tests for this use-case and a demo that didn't work before and works now. — You are receiving this because you were mentioned. Reply to this email directly, view it on GitHub<#800?email_source=notifications&email_token=AAHODI4P6OWXLABXTTPSZOLQOHPA3A5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBCA45I#issuecomment-541331061>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODI3REYBQRRH4343S45DQOHPA3ANCNFSM4I7ODUCQ>.

> @amergey This changes the behavior of the authenticator. It needs tests for the use-case where `getServletPath` returns a `null`, tests for any updated logical path, and ideally a demo that didn't work before and works now.

Unless I missed something current behavior is not consistent between negotiate and form authentication so it needs to be fixed anyway. 

- If negotiate authentication behavior is expected one, then code should be

```
else if (negotiateCheck) {
            if (!authorizationHeader.isNull()) {
                return this.negotiate(request, response, authorizationHeader);
            }
            this.log.debug("authorization required");
            this.sendUnauthorized(response);
            return false;
        } else if (securityCheck) {
            return this.redirectTo(request, response, request.getServletPath());
        }

```

- If form authentication behavior is expected then code should be

```
else if (negotiateCheck) {
            if (!authorizationHeader.isNull()) {
              final boolean result = this.negotiate(request, response, authorizationHeader);
              if (result) {
                this.redirectTo(request, response, request.getServletPath());
              } else {
                this.redirectTo(request, response, loginConfig.getErrorPage());
              }
              return result;
            }
            this.log.debug("authorization required");
            this.sendUnauthorized(response);
            return false;
        } else if (securityCheck) {
            final boolean postResult = this.post(request, response);
            if (postResult) {
                this.redirectTo(request, response, request.getServletPath());
            } else {
                this.redirectTo(request, response, loginConfig.getErrorPage());
            }
            return postResult;
        }
```
Even if you think the issue raised is not something to change (fix), or there is a better thing to do, having a consistent behavior is a must have and probably the more important point here.

dblock commented:
I am all for consistent behavior, but I also want to understand why a change is needed and to see tests to avoid breaking something new behavior fixes.

amergey commented:
As I said before, MixedAuthenticator is already covered by a test case (waffle.apache.MixedAuthenticatorTests). 
The redirect part where the issue occurs is not tested because redirect is handled by Tomcat Servlet implementation. Waffle tests are not executed with tomcat implementation but with Mocked Http Request and Reponse (which is totally fine as you do not want to test tomcat implementation here).

That's said in my last commit I have updated the demo source code to show you the issue, if you build it and you try to browse 
**http://localhost:8080/waffle-mixed-1.9.0/test** both negociate and form authentication are working
but if you replace waffle jars with the one from the latest release you will see negociate is working (because no sendRedirect) but form authentication is failing with the IllegalStateException.

If you want to avoid behavior not covered by test then you need to remove redirectTo method from MixedAuthenticator (but then with the issue of not showing the error page in case of failure), there is no easy way to cover sendRedirect behavior in a test case.

hazendaz commented:
Thanks! I'll look at this more over the weekend.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arnaud Mergey <notifications@github.com>
Sent: Thursday, October 17, 2019 11:09:27 AM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


As I said before, MixedAuthenticator is already covered by a test case (waffle.apache.MixedAuthenticatorTests).
The redirect part where the issue occurs is not tested because redirect is handled by Tomcat Servlet implementation. Waffle tests are not executed with tomcat implementation but with Mocked Http Request and Reponse (which is totally fine as you do not want to test tomcat implementation here).

That's said in my last commit I have updated the demo source code to show you the issue, if you build it and you try to browse
http://localhost:8080/waffle-mixed-1.9.0/test both negociate and form authentication are working
but if you replace waffle jars with the one from the latest release you will see negociate is working (because no sendRedirect) but form authentication is failing with the IllegalStateException.

If you want to avoid behavior not covered by test then you need to remove redirectTo method from MixedAuthenticator (but then with the issue of not showing the error page in case of failure), there is no easy way to cover sendRedirect behavior in a test case.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI6LJHWFMMYGLGUCJETQPB52NA5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEBQOH7I#issuecomment-543220733>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODI266TLPTSRLNYKGWTDQPB52NANCNFSM4I7ODUCQ>.

amergey commented:
I could also have demoed a possible issue caused by the inconsistency between negociate and form authentication. With some servlet mappings configuration you can end up with an application where when you authenticate through form, you end up in a different page than through negociate, because form is redirecting to servlet path, and negociate is not redirecting. This is a minor issue compared to the IllegalArgumentException, my proposal is fixing that as well

hazendaz commented:
@amergey I think a need a bit more time to digest this.  I did a couple of cleanup items on this PR.  I restored 'space' formatting which is preferred over 'tab' usage so we are now only looking at the changes you have made directly.

As to why negoticate differs.  I'm pretty sure that is because during negotiation the browser and the container are talking with each other not our direct calls.  Thus I think there is going to be some behavior change here that is unexpected.

The the security change, what I'm gathering here is that you intend to leave the page where its at and only change when necessary to go back to logon page.

In both instances, I need to think about this more.  Truth be told, I rarely use anything but the 'waffle-jna' module piece which 99% of the time I feel is all that is necessary anyways.  I get making the entire container being enabled by this but reality these days is that everything is going spring boot, drop wizard, or singular anyways which sort of deminishes the larger value this does provide.  That said, that also means I don't have a general setup at all to really test at this level outside the demo which I'll get approved through anyways.  So I'll need a bit more time to test.  I cannot at the moment make a judgement call as this being right or wrong in any way.  My gut tells me there is some consideration here due to how underlying processing works.

The bigger thing I'm having trouble with is various mention throughout this thread of not being consistent with other portions of code but I'm having trouble pinpointing the inconsistency to start with. Could you do a few more things here?  First, please directly pinpoint the line of code that you feel is inconsisent and how this change makes it more consistent?  Second, can you show error results you are receiving with the code as-is?

amergey commented:
Regarding issues:

`this.redirectTo(request, response, request.getServletPath()); `
is failing when request.getServletPath is returning "", I have updated the demo to illustrate that, with servlet mapping like this /* it can happened.

other issue, because of 
`return this.negotiate(request, response, authorizationHeader);`
if negotiate authentication is failing, a blank page is displayed (this has been fixed for form only by redirecting to error page) 

I propose this changes mainly to make negotiate and form authentication behavior consistant
Bonuses are :
- it also fixes redirect failure after form authentication when request.getServletPath return null or empty string
- it also fixes blank display when negotiate authentication is failing



- behavior consistent

hazendaz commented:
I'm not able to demonstrate any issue before or after this change to show what exactly is occurring.  I'm also not clear how servlet path is ever null.  Tomcat does not define that possibility.  It does with other elements in the request but not that one.

'test.jsp' is just a junk non existent page.  The global catch all addition of waffle servlet simply is catching that and overriding the default behaviour.  I see no reason for that.

In order to move forwards here, I would need a basic complete bullet point walk through on how to make said issue occur.

hazendaz commented:
I should note: tomcat specifically documents that servletPath will be an empty string if there is nothign to add.

amergey commented:
Note described issue is occuring when servlet path is returning empty string as well.

> _**I'm not able to demonstrate any issue before** or after **this change**_
@hazendaz  

After october 14 comment, I added to the pull request a change on demo that is showing the IllegalStateException issue, do you mean you have try the updated demo and you do not see this issue ? I am pretty sure it is occuring consistently, I am still reproducing it. 
You can think the consistency issue I raised here is not an issue, because there is no failure, just "inconsistent navigation path" comming from sso authentication vs form authentication, but the IllegalStateException occuring when servlet path is returning null or empty string is clearly an issue, as an exception is raised instead

amergey commented:
I do not get this 
"'test.jsp' is just a junk non existent page" 
the demo I updated is not going to any non existing test.jsp it is going to "/test" which is supposed to display waffle.servlet.WaffleInfoServlet, thanks to /* mapping, so it should not display a blank page but supposed to display what  WaffleInfoServlet is supposed to display

hazendaz commented:
Which version of tomcat were you seeing this behaviour?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arnaud Mergey <notifications@github.com>
Sent: Friday, December 27, 2019 3:14:08 PM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


I do not get this
"'test.jsp' is just a junk non existent page"
the demo I updated is not going to any non existing test.jsp it is going to "/test" which is supposed to display waffle.servlet.WaffleInfoServlet, thanks to /* mapping, so it should not display a blank page but supposed to display what WaffleInfoServlet is supposed to display

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI5DCL4IPPM6DAYBDQ3Q2ZOZBA5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXWUOQ#issuecomment-569338426>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODI23XNCK6SM5L7TV2CLQ2ZOZBANCNFSM4I7ODUCQ>.

hazendaz commented:
Which browser?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Jeremy Landis <jeremylandis@hotmail.com>
Sent: Friday, December 27, 2019 3:30:28 PM
To: Waffle/waffle <reply@reply.github.com>; Waffle/waffle <waffle@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)

Which version of tomcat were you seeing this behaviour?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arnaud Mergey <notifications@github.com>
Sent: Friday, December 27, 2019 3:14:08 PM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


I do not get this
"'test.jsp' is just a junk non existent page"
the demo I updated is not going to any non existing test.jsp it is going to "/test" which is supposed to display waffle.servlet.WaffleInfoServlet, thanks to /* mapping, so it should not display a blank page but supposed to display what WaffleInfoServlet is supposed to display

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI5DCL4IPPM6DAYBDQ3Q2ZOZBA5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHXWUOQ#issuecomment-569338426>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODI23XNCK6SM5L7TV2CLQ2ZOZBANCNFSM4I7ODUCQ>.

amergey commented:
This is not an issue related to browser, this occurs with tomcat 8.5 and 9 as well. 
Please read carrefully my previous comments, and changes I made to the code in my pull request, all explanations should be already there

More the mine from october 22 

> _Regarding issues:

this.redirectTo(request, response, request.getServletPath());
is throwing exception when request.getServletPath is returning a path not starting with "/", I have updated the demo to illustrate how it can happened, for example with servlet mapping /*.
_

The updated demo demonstrate this issue because, the way I update the demo, when you try to navigate to any page, the login.jsp form is displayed, then when you try to authenticate, waffle is redirecting to test?j_security_check if you choose form authentication . Thanks to /* servlet mapping, test is supposed to redirect to waffle.servlet.WaffleInfoServlet, but it fails with illegalstate exception because getServletPath returns empty string.
Note that if you change the demo and replace /* with /test for servlet mapping, it will then succeed because getServletPath would return /test 

The key to understand issues, is to understand waffle is doing redirect using **getServletPath** which itself is depending of the way servlet mapping is configured, this is why it seems cleaner to get rid of it. 

Of course to reproduce the issue, you need to take only  the update demo web.xml with current waffle code, if you build the demo from my pull request, as I fixed waffle in the same time you will not reproduce the issue.

hazendaz commented:
Hmmm.  I cannot reproduce the issue with current code.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arnaud Mergey <notifications@github.com>
Sent: Saturday, December 28, 2019 4:22:41 AM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


This is not an issue related to browser, this occurs with tomcat 8.5 and 9 as well.
Please read carrefully my previous comments, and changes I made to the code in my pull request, all explanations should be already there

More the mine from october 22

_Regarding issues:

this.redirectTo(request, response, request.getServletPath());
is throwing exception when request.getServletPath is returning a path not starting with "/", I have updated the demo to illustrate how it can happened, for example with servlet mapping /*.
_

The updated demo demonstrate this issue because, the way I update the demo, when you try to navigate to any page, the login.jsp form is displayed, then when you try to authenticate, waffle is redirecting to test?j_security_check if you choose form authentication . Thanks to /* servlet mapping, test is supposed to redirect to waffle.servlet.WaffleInfoServlet, but it fails with illegalstate exception because getServletPath returns empty string.
Note that if you change the demo and replace /* with /test for servlet mapping, it will then succeed because getServletPath would return /test

The key to understand issues, is to understand waffle is doing redirect using getServletPath which itself is depending of the way servlet mapping is configured, this is why it seems cleaner to get rid of it.

Of course to reproduce the issue, you need to take only the update demo web.xml with current waffle code, if you build the demo from my pull request, as I fixed waffle in the same time you will not reproduce the issue.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI4RJ3525XN3FAUIYUTQ24LGDA5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYF62Q#issuecomment-569401194>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODIZNIOUCXBVATQ5R7ODQ24LGDANCNFSM4I7ODUCQ>.

hazendaz commented:
Is there anything different with your tomcat installation? Can I see the full stack trace you receive?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Jeremy Landis <jeremylandis@hotmail.com>
Sent: Saturday, December 28, 2019 1:18:57 PM
To: Waffle/waffle <reply@reply.github.com>; Waffle/waffle <waffle@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)

Hmmm.  I cannot reproduce the issue with current code.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arnaud Mergey <notifications@github.com>
Sent: Saturday, December 28, 2019 4:22:41 AM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


This is not an issue related to browser, this occurs with tomcat 8.5 and 9 as well.
Please read carrefully my previous comments, and changes I made to the code in my pull request, all explanations should be already there

More the mine from october 22

_Regarding issues:

this.redirectTo(request, response, request.getServletPath());
is throwing exception when request.getServletPath is returning a path not starting with "/", I have updated the demo to illustrate how it can happened, for example with servlet mapping /*.
_

The updated demo demonstrate this issue because, the way I update the demo, when you try to navigate to any page, the login.jsp form is displayed, then when you try to authenticate, waffle is redirecting to test?j_security_check if you choose form authentication . Thanks to /* servlet mapping, test is supposed to redirect to waffle.servlet.WaffleInfoServlet, but it fails with illegalstate exception because getServletPath returns empty string.
Note that if you change the demo and replace /* with /test for servlet mapping, it will then succeed because getServletPath would return /test

The key to understand issues, is to understand waffle is doing redirect using getServletPath which itself is depending of the way servlet mapping is configured, this is why it seems cleaner to get rid of it.

Of course to reproduce the issue, you need to take only the update demo web.xml with current waffle code, if you build the demo from my pull request, as I fixed waffle in the same time you will not reproduce the issue.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI4RJ3525XN3FAUIYUTQ24LGDA5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYF62Q#issuecomment-569401194>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODIZNIOUCXBVATQ5R7ODQ24LGDANCNFSM4I7ODUCQ>.

hazendaz commented:
Can you outline everything you setup to run this?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Jeremy Landis <jeremylandis@hotmail.com>
Sent: Saturday, December 28, 2019 1:22:08 PM
To: Waffle/waffle <reply@reply.github.com>; Waffle/waffle <waffle@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)

Is there anything different with your tomcat installation? Can I see the full stack trace you receive?

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Jeremy Landis <jeremylandis@hotmail.com>
Sent: Saturday, December 28, 2019 1:18:57 PM
To: Waffle/waffle <reply@reply.github.com>; Waffle/waffle <waffle@noreply.github.com>
Cc: Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)

Hmmm.  I cannot reproduce the issue with current code.

Get Outlook for Android<https://aka.ms/ghei36>

________________________________
From: Arnaud Mergey <notifications@github.com>
Sent: Saturday, December 28, 2019 4:22:41 AM
To: Waffle/waffle <waffle@noreply.github.com>
Cc: Jeremy Landis <jeremylandis@hotmail.com>; Mention <mention@noreply.github.com>
Subject: Re: [Waffle/waffle] Fix form authentication in tomcat MixedAuthenticator valve (#800)


This is not an issue related to browser, this occurs with tomcat 8.5 and 9 as well.
Please read carrefully my previous comments, and changes I made to the code in my pull request, all explanations should be already there

More the mine from october 22

_Regarding issues:

this.redirectTo(request, response, request.getServletPath());
is throwing exception when request.getServletPath is returning a path not starting with "/", I have updated the demo to illustrate how it can happened, for example with servlet mapping /*.
_

The updated demo demonstrate this issue because, the way I update the demo, when you try to navigate to any page, the login.jsp form is displayed, then when you try to authenticate, waffle is redirecting to test?j_security_check if you choose form authentication . Thanks to /* servlet mapping, test is supposed to redirect to waffle.servlet.WaffleInfoServlet, but it fails with illegalstate exception because getServletPath returns empty string.
Note that if you change the demo and replace /* with /test for servlet mapping, it will then succeed because getServletPath would return /test

The key to understand issues, is to understand waffle is doing redirect using getServletPath which itself is depending of the way servlet mapping is configured, this is why it seems cleaner to get rid of it.

Of course to reproduce the issue, you need to take only the update demo web.xml with current waffle code, if you build the demo from my pull request, as I fixed waffle in the same time you will not reproduce the issue.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Waffle/waffle/pull/800?email_source=notifications&email_token=AAHODI4RJ3525XN3FAUIYUTQ24LGDA5CNFSM4I7ODUC2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEHYF62Q#issuecomment-569401194>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAHODIZNIOUCXBVATQ5R7ODQ24LGDANCNFSM4I7ODUCQ>.

amergey commented:
No particular setup, you should be able to reproduce with current waffle code, as I said the issue occurs when getServletPath is returning empty string, and the way to have a getServletPath returning empty string is to navigate to a servlet with /* as servlet mapping. Have you tried to debug in order to check what getServletPath is returning in your case ?

amergey commented:
Note the issue only occurs with MixedAuthenticator valve and only with form authentication (?j_security_check), sso is working without errors even with current code (j_negociate_check)

amergey commented:
"As a proof of my claim" according to tomcat javadoc 
https://tomcat.apache.org/tomcat-8.5-doc/servletapi/javax/servlet/http/HttpServletRequest.html#getServletPath()

**This method will return an empty string ("") if the servlet used to process this request was matched using the "/*" pattern.**

 and ServletContext javadoc https://docs.oracle.com/javaee/6/api/javax/servlet/ServletContext.html#getRequestDispatcher(java.lang.String)
It is clearly said that **The pathname must begin with a /**, 

(https://github.com/Waffle/waffle/blob/master/Source/JNA/waffle-tomcat85/src/main/java/waffle/apache/MixedAuthenticator.java#L275)

hazendaz commented:
Here is how things are working to re-baseline the progress here.

-- Current Code as-is --

Environment
- Tomcat 9.0.30
- Lasest Waffle 'master' build (2.1.1-SNAPSHOT)
- My tomcat uses logback and is more tilted towards EE but otherwise should behave mostly the same. You can get a copy of this from my github under 'hazendaz/tomcat9'.  It contains logback-core 1.2.3
- tomcat/conf -> added entry into my logback.xml to pick up debug of waffle (<logger name="waffle.apache" level="DEBUG" />).  This could be adjusted to just pick up anything waffle but this is enough to demonstrate things here.
- tomcat/lib -> added caffeine 2.8.0, jna-5.5.0, jna-platform-5.5.0, logback-classic 1.2.3, slf4j-api-1.7.30, waffle-jna-2.1.1-SNAPSHOT, and waffle-tomcat9-2.1.1-SNAPSHOT
- tomcat/web -> added waffle-mixed.war (2.1.1-SNAPSHOT)

Security Check Path
=============
- If invalid, properly reports error page with 'An error has occurred."
- If valid, works as expected on the view resulting in login information being deplayed.  Under the logs it gets this error "Unable to add HTTP headers since response is already committed on entry to the HTTP header security Filter"

Negotiation Path
===========
- If 'cancel' selected, reports 'Access Denied'
- If invalid, it will continually renegotiate with user
- If valid, works as expected resulting in login information being displayed

At this point, there is a valid concern with Security Check Path when it works in general.  The PR itself would appear to resolve that by no longer trying to redirect to the servletPath.  Thus my next test is to just address that piece by itself to see if behaviour works without an exception in the logs.

Where I stand overall.  I believe the negotiation changes are completely unnecessary.  I don't agree with the consistency issue in that Security check path and negotation are really two entirely different things and them being aligned to act the same would change the behaviour of negotiate as this PR is doing.  However, I'm going on assumption here based on the baseline test.  I will further test it but expect that the 'cancel' option will result in stopping negotiation and displaying error page.  This is not typical behaviour I've seen in any security piece when using negotiation to date.  Its not to say that is not valid, but generaly most other frameworks behave in this way from my experience.  I think (and again will test), that the change will make it 2 paths (success or failure) regardless of the 2 overall security paths.  That does make it consistent.  However, I don't believe that is the current issue and probably should be handled on a separate PR if decided that is worth having.

Also to note, the current issue is suggesting we are dealing with a null pointer exception.  That is not the case as it is simply empty string from tomcat as documented from at least tomcat 5.5 until now based on reviewing tomcat javadocs.  So we at least need to rename what this is actually fixing.

I'm not sure we additionally want to change how the demo works.  As there is a fundemental problem here that I now see, that behaviour change to the demo is rather drastic.  I think for now I continue on testing applying bits and pieces of the logic from this PR to see how things behave.  As this improves, I'll likely merge in smaller pieces then toss it back over for @amergey to take a look at.

Hang off for now in responding back.  Let me finish more testing here later tonight and get something more formal of my findings past this point (ie formal testing is as-is code without any modifications to the demo).

hazendaz commented:
@amergey ok I'm done testing.  I need a few things from you then I can merge this in.

1 -> Please update the change log to reflect the 2 different items being addressed here.  Here is how I see it.  Feel free to adjust as needed in order to make the most sense to other users.

- Security Check in tomcat valves performs a redirect to servletPath when successful.  This causes an underlying error and is not required to finish the chain.  This redirect has been removed.
- Negotiate Check in tomcat valvles performs half the necessary negotiation which is resulting in popup to log into windows.  By catching the negotiation result and forcing a redirect to error page as intended (simplar to security check), the browser and tomcat are able to successfully negotiate the communication without unnecessary popup to the user.

2 -> Remove the changes applied for the demo.  Ultimately I found that not as useful as it changed too much behaviour as this was eventually determined to be an issue with the demo as-is.  If you would like to demonstrate this as you have and feel it is necessary, I think possibly a secondary demo should be used.

3 -> Actual code changes are good as-is.

Once issues are addressed and this merged, I can move to formally release this.  I will bump the release to 2.2.0 for this change.  Please reflect that in the change log :)  Also to note, I had just ran a release with 2.1.1 thus this is a perfect spot for behaviour change for the better.

@dblock Any further thoughts here?  Anything I might have missed?

amergey commented:
I have updated changelog and reverted demo changes, I initially changed demo because it was requested to me to show a way to reproduce the issue, but I agree with you this demo changes is not useful here.

hazendaz commented:
@amergey Awesome!  Thanks.  I'll push a release now.  It should be available in a few hours on central.

hazendaz commented:
@amergey I'll have to come back to releasing this tomorrow.  I forgot our tests cases don't run on travis (linux...).  There are test failures now.  I need to review them.  I'll get back to this likely tomorrow evening.  Thanks.

hazendaz commented:
The test case is for tomcat7 only.  We were not previously running redirect so login was set as null.  Now that matters.  No other tests needed fixed.  While I have finished that and sucessfully retested, I'm now reviewing sonar related issue around security problems and I discovered the spring boot modules were broken as the original design did not properly define the JNA version at the modules and only did so in demo which was itself half broken.  So now I have ensured that is clear for all. I believe a number of folks were running into that as well.  As a result of all this, I'm likely to hang off until sometime this weekend so I can get a few more pushes through sonar and try to reduce the issues it is noting.  I further want to look again at a few issues that seem common going back years with IE posting 0 size resulting in IE terminating the negotiation.  I think based on all the reports that might be easily fixed but need to review those more.
